### PR TITLE
Dynamic repository location

### PR DESCRIPTION
**Pull request by JakeGinnivan**

### CODE REVIEW

File: GitVersionExe/GitPreparer.cs [hunk 1]
```diff
@@ -20,31 +21,76 @@ public bool IsDynamicGitRepository
...
+            var targetPath = CalculateTemporaryRepositoryPath(arguments.TargetUrl);
+            DynamicGitRepositoryPath = CreateDynamicRepository(targetPath, arguments.Authentication, arguments.TargetUrl, arguments.TargetBranch);
+        }
+
+        string CalculateTemporaryRepositoryPath(string targetUrl)
+        {
+            var userTemp = Path.GetTempPath();
+            var repositoryName = targetUrl.EndsWith(".git") ?
+                targetUrl.Split('/', '\\').Last().Replace(".git", string.Empty) :
+                targetUrl.Split('/', '\\').Last();
```
Comment (eggapauli) [inline GitVersionExe/GitPreparer.cs]: I would either use [@GeertvanHorrik 's approach](https://github.com/ParticularLabs/GitVersion/issues/398#issuecomment-84414898) or:

    var repositoryName = new Uri(targetUrl).Segments.Last();
    repositoryName = repositoryName.EndsWith(".git")
        ? Path.GetFileNameWithoutExtension(repositoryName)
        : repositoryName;

File: GitVersionExe/GitPreparer.cs [hunk 2]
```diff
@@ -20,31 +21,76 @@ public bool IsDynamicGitRepository
...
+            var possiblePath = Path.Combine(userTemp, repositoryName);
+
+            // Verify that the existing directory is ok for us to use
```
Comment (eggapauli) [inline GitVersionExe/GitPreparer.cs]: Is this still necessary when updating existing repos?

**Comment by JakeGinnivan** [comment]
@eggapauli @GeertvanHorrik ok, I think I have fixed the issues pointed out.

File: GitVersionExe/GitPreparer.cs [hunk 3]
```diff
@@ -20,31 +21,74 @@ public bool IsDynamicGitRepository
...
 
-        string GetGitInfoFromUrl()
+        static bool GitRepoHasMatchingRemote(string possiblePath, string targetUrl)
         {
-            var gitRootDirectory = Path.Combine(arguments.TargetPath, "_dynamicrepository");
-            var gitDirectory = Path.Combine(gitRootDirectory, ".git");
-            if (Directory.Exists(gitRootDirectory))
+            try
+            {
+                using (var repository = new Repository(possiblePath))
```
Comment (GeertvanHorrik) [inline GitVersionExe/GitPreparer.cs]: Not sure (haven't looked into the code), but shouldn't we do a fetch first to check if there are new remotes (for example, for PR)?

File: GitVersionExe/GitPreparer.cs [hunk 4]
```diff
@@ -20,31 +21,74 @@ public bool IsDynamicGitRepository
...
-                DeleteHelper.DeleteGitRepository(gitRootDirectory);
+            return GitDirFinder.TreeWalkForDotGitDir(arguments.TargetPath);
+        }
+
+        static string CreateDynamicRepository(string targetPath, Authentication authentication, string repositoryUrl, string targetBranch)
+        {
+            var gitDirectory = Path.Combine(targetPath, ".git");
+            if (Directory.Exists(targetPath))
+            {
+                Logger.WriteInfo("Git repository already exists at {0}, skipping clone");
```
Comment (GeertvanHorrik) [inline GitVersionExe/GitPreparer.cs]: Clone should be skipped, but shouldn't we be doing a fetch / pull?

Comment (JakeGinnivan) [inline GitVersionExe/GitPreparer.cs]: The normalise git repo step later will fetch/normalise for us? This side of GitVersion I'm not too strong on though.

Sent from my Windows Phone
________________________________

File: GitVersionExe/GitPreparer.cs [hunk 5]
```diff
@@ -20,31 +21,74 @@ public bool IsDynamicGitRepository
...
         {
-            var gitPath = arguments.TargetPath;
+            if (string.IsNullOrWhiteSpace(arguments.TargetUrl)) return;
 
-            if (!string.IsNullOrWhiteSpace(arguments.TargetUrl))
+            var targetPath = CalculateTemporaryRepositoryPath(arguments.TargetUrl);
+            DynamicGitRepositoryPath = CreateDynamicRepository(targetPath, arguments.Authentication, arguments.TargetUrl, arguments.TargetBranch);
+        }
+
+        string CalculateTemporaryRepositoryPath(string targetUrl)
```
Comment (eggapauli) [inline GitVersionExe/GitPreparer.cs]: I would still love to create the dynamic repo in a custom location, i.e. set via a cmd-line argument.

File: GitVersionExe/GitPreparer.cs [hunk 6]
```diff
@@ -20,31 +21,74 @@ public bool IsDynamicGitRepository
...
+        {
+            var userTemp = Path.GetTempPath();
+            var repositoryName = targetUrl.Split('/', '\\').Last().Replace(".git", string.Empty);
+            var possiblePath = Path.Combine(userTemp, repositoryName);
+
```
Comment (eggapauli) [inline GitVersionExe/GitPreparer.cs]: Ahh I just saw that it's created in tmp. That's perfectly fine I guess.

**Pull request merged**